### PR TITLE
add coroutine behavior for generators

### PR DESCRIPTION
**Pull request by WarriorOfWire**
coroutines don't have __next__; they also call themselves coroutines.
This does not change the fact that `async def` methods are generators,
but it does make them behave more like CPython.

This also turns on `async` in mpy-cross so tests will pass, though the exception thrown by the mpy-cross byte code is not very helpful.

### CODE REVIEW

File: tests/basics/async_coroutine.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+try:
+    f()  # Should not crash
+except Exception as e:
+    print('failed to invoke')
+
+try:
+    next(f())
+    print('This should fail because async def returns a coroutine, and next() is not allowed')
+except Exception as e:
```
Comment (tannewt) [inline tests/basics/async_coroutine.py:12]: Want to make this TypeError explicitly?

Comment (WarriorOfWire) [reply tests/basics/async_coroutine.py:12]: I do, but if you catch TypeError only, the test run that does
```
MICROPY_CPYTHON3=python3.8 MICROPY_MICROPYTHON=../ports/unix/micropython_coverage ./run-tests -j1 --via-mpy -d basics float
```
fails.

**Comment by tannewt** [comment]
What do you think about async only being enabled on non-m0?

**Comment by dhalbert** [comment]
> What do you think about async only being enabled on non-m0?

I could imagine async being used for simple multi-task timing projects even on a Trinket M0, say. I think a simple task library would fit.

Is the problem with the bast pro mini that it has more pins than other small boards like trinket and gemma?

**Comment by dhalbert** [comment]
While looking at #3190, I realized that both `py/gc.o` and `py/vm.o` are compiled `-O3` by default. We have turned that off for GC on small builds (`SUPEROPT_GC =0`), but not for `vm.o`. Compiling `vm.o` as `-O2` instead of `-O3` gains back 416 bytes on a `trinket_m0` build.

**Comment by tannewt** [comment]
> > What do you think about async only being enabled on non-m0?
> 
> I could imagine async being used for simple multi-task timing projects even on a Trinket M0, say. I think a simple task library would fit.

Sure, but that doesn't mean we need it enabled. There are lots of things that could be useful on an M0.

> Is the problem with the bast pro mini that it has more pins than other small boards like trinket and gemma?

Yes, I think so.

> While looking at #3190, I realized that both `py/gc.o` and `py/vm.o` are compiled `-O3` by default. We have turned that off for GC on small builds (`SUPEROPT_GC =0`), but not for `vm.o`. Compiling `vm.o` as `-O2` instead of `-O3` gains back 416 bytes on a `trinket_m0` build.

Nice! That'll do. How much do we gain with `-Os`? I wonder what the speed loss is. GCC10 will also make builds smaller.

**Comment by dhalbert** [comment]
> > I could imagine async being used for simple multi-task timing projects even on a Trinket M0, say. I think a simple task library would fit.
> 
> Sure, but that doesn't mean we need it enabled. There are lots of things that could be useful on an M0.

I was just thinking that if it was _really_ easy to use, it could be used in simple blinky and NeoPixel programs, etc.

> > While looking at #3190, I realized that both `py/gc.o` and `py/vm.o` are compiled `-O3` by default. We have turned that off for GC on small builds (`SUPEROPT_GC =0`), but not for `vm.o`. Compiling `vm.o` as `-O2` instead of `-O3` gains back 416 bytes on a `trinket_m0` build.
> 
> Nice! That'll do. How much do we gain with `-Os`? I wonder what the speed loss is. GCC10 will also make builds smaller.

Well, that's weird: `-Os` should be the best, but it's not:

`CSUPEROPT` (affects `vm.o` build)  | bytes free on `trinket_m0` build
---|---
`-Os` | 2292
`-O2` | 2412
`-O3` | 1996

**Comment by WarriorOfWire** [comment]
> What do you think about async only being enabled on non-m0?

Async is useful everywhere; right now it's nowhere.  There's no great loss if it's not available on m0 at the outset.

It might be a problem in the future, as it's a core language feature, but on the short-to-medium term it seems fine.  Right now I'm happy to disable any target that does not easily build (given that there is a pretty healthy crop of boards that _do_ build).  If there is demand on those tiny boards the problems can be solved.

**Comment by WarriorOfWire** [comment]
Let's try a rebase and see what shakes loose

**Comment by WarriorOfWire** [comment]
@tannewt okay, this passes after a rebase so the can is kicked one commit further down the road.

**Pull request merged**